### PR TITLE
ABI generation

### PR DESCRIPTION
**Pull request by itegulov**
This is still pretty raw, but sharing for visibility.

Instructions for running:
1. Clone https://github.com/near/near-abi-js to a sibling directory and run `yarn install && yarn build` (needed because it is not released yet)
2. Clone https://github.com/itegulov/typescript-json-schema to a sibling directory and run `yarn install && yarn build` (needed because I had to expose some API that makes life much much easier).
3. Run `yarn build:counter-ts` in `examples` and see what happens :eyes:

### CODE REVIEW

File: src/cli/abi.ts
Comment (itegulov) [inline src/cli/abi.ts]: We can get rid of the preceding CLI step that does typechecking since this piece does effectively the same without having to instantiate `tsc` twice. But up to you guys.

File: package.json
Comment (volovyks) [inline package.json]: Why do we need this script?

File: package.json
Comment (volovyks) [inline package.json]: @itegulov 
Is it a separate crate in Rust? If yes, looks like we will need to move to the monorepo arch @ailisp.

File: src/version.ts
Comment (volovyks) [inline src/version.ts]: Probably we can get it from package.json.

File: src/cli/abi.ts
Comment (volovyks) [reply src/cli/abi.ts]: It is still unclear to me if IBI is a separate package. If it is not - then we can do a TS check once.

File: package.json
Comment (itegulov) [reply package.json]: It is. Basically an unopinionated set of models that people can use to parse/work with ABIs: https://github.com/near/near-abi-rs. If you don't think it will be useful as its own library, I can move it here for sure.

File: src/version.ts
Comment (itegulov) [reply src/version.ts]: This relates to what I was talking about in Slack. If I import package.json, then it messes up how sources are placed in the `lib` folder. For transparency this is the Slack message:

>  I figured out why. Apparently if you import any file outside of src, then tsc will maintain the src directory for you. So when I did this:
`import sdkPackage from '../../package.json';`
It made so that package.json was copied to `lib` and hence the source code had to go into `src`. Which makes sense, but came as a bit of surprise.

File: package.json
Comment (itegulov) [reply package.json]: Basically a roundabout way of getting library version without having to import package.json. Definitely hacky, so feel free to suggest an alternative!

File: src/cli/abi.ts
Comment (itegulov) [reply src/cli/abi.ts]: IBI?

File: src/version.ts
Comment (miraclx) [reply src/version.ts]: How about `JSON.parse(path.join(__dirname, "..", "package.json"))`?

File: package.json
Comment (ailisp) [reply package.json]: Since it's similarly organized as rust, as a separate looks good

File: src/cli/abi.ts
Comment (ailisp) [inline src/cli/abi.ts:6]: It looks like this generated version.js isn't necessary, since you have `parseMetadata` below, which still load package.json?

File: src/cli/abi.ts
Comment (ailisp) [inline src/cli/abi.ts]: So these overrides options load in basePath/tsconfig.json?

File: src/cli/abi.ts
Comment (ailisp) [inline src/cli/abi.ts]:
```suggestion
                    throw Error("NEAR function cannot be init, call and view at the same time");
```

**Comment by ailisp** [comment]
Generally your implementation and comprehensive validation looks great!

File: src/cli/abi.ts
Comment (ailisp) [inline src/cli/abi.ts:102]: we may also want to ensure it's under a parent node of `@NearBindgen`

File: src/cli/abi.ts
Comment (itegulov) [reply src/cli/abi.ts:6]: `parseMetadata` actually loads the package file supplied by the user (i.e. contract's package.json). But I also need to parse the near-sdk-js package.json. We can go with what @miraclx suggested in the other thread though if that looks good to you.

File: src/cli/abi.ts
Comment (itegulov) [reply src/cli/abi.ts]: So I tried to use contract's `tsconfig.json`, but then it picks up all TS files in this project, not just the one you directly specify (e.g. all examples under `/examples`, not just `counter-ts`). I will play around with this a bit more and see what can be done.

File: src/version.ts
Comment (itegulov) [reply src/version.ts]: @miraclx couldn't use your snippet as `__dirname` is not supported in ES modules. Went with `JSON.parse(fs.readFileSync(path.join(fileURLToPath(import.meta.url), "..", "..", "package.json"), 'utf-8'));` as the next best option.

File: src/cli/abi.ts
Comment (itegulov) [reply src/cli/abi.ts]: I think I figured out the right way to deal with this, PTAL.

While we are at this, how do you think the path to `package.json` and `tsconfig.json` should be specified? I have just added a 3rd and 4th argument to the CLI command, but that's probably not what we want.

**Comment by itegulov** [comment]
I am working on stabilizing this and I have a couple of things that I am not sure how to go about.

1. Should this be an opt-in feature for now? How do you guys imagine the CLI interface for ABI generation should look like? Do we also provide a separate command `abi` that just generates ABI without building the contract? I can match what we have in Rust, but not sure if that makes sense as the build tools are quite different, so your call.
2. I have released https://github.com/itegulov/typescript-json-schema as `near-typescript-json-schema` for now just to unblock this PR, but not sure how you would like to handle this fork. We have a similar situation with https://github.com/near/schemafy/, so we just moved the fork under `near` org and are maintaining it ourselves. Ideally we should discuss this with the original repo maintainers and see if they would be open to exposing the APIs that we need, but in my experience open-source work is pretty slow, so I wouldn't count on it short-term (we are still waiting for schemafy maintainers' response for instance).

So far most examples work, only NFT is failing because `bigint` is apparently not supported by `typescript-json-schema`. There is [this PR](https://github.com/YousefED/typescript-json-schema/pull/433) that was never merged though, so it seems like I can easily add support in our fork if you don't have any objections.

File: src/cli/abi.ts
Comment (volovyks) [reply src/cli/abi.ts]: Sorry, "ABI"

**Comment by volovyks** [comment]
@itegulov our build command now consists of several steps. I think ABI can become one of them. It should be easy to have as a separate command as well. I do not think that we need to keep it as a separate project.

File: src/version.ts
Comment (miraclx) [reply src/version.ts]: PS: this is equivalent to

`JSON.parse(fs.readFileSync(fileURLToPath(new URL("../package.json", import.meta.url)), "utf-8"))`. 😄

Comment (itegulov) [reply src/version.ts]: This wouldn't work on Windows though, right?

Comment (miraclx) [reply src/version.ts]: Tested, it works. `fileUrlToPath` makes the conversion.

Comment (itegulov) [reply src/version.ts]: Ah, nice! I will change this over then

File: examples/src/parking-lot.ts
Comment (ailisp) [inline examples/src/parking-lot.ts:70]: will we get warnings (or configure tsconfig in a way) when method doesn't have return type?

**Review by ailisp — APPROVED** [review]
Nice work! It's very solid implementation! Also thank you for fixing a few signatures!

**Review by volovyks — APPROVED** [review]
Great work @itegulov!
Merging now, but do you think it's possible to test ABI somehow?

**Pull request merged**
